cmd/kubelet/app/server.go

```diff
@@ -513,6 +537,47 @@ func run(s *options.KubeletServer, kubeDeps *kubelet.KubeletDeps) (err error) {
 	return nil
 }
 
+func initializeCertificateManager(certDirectory string, certRotation int32, nodeName types.NodeName, certData []byte, keyData []byte) (certificate.Manager, error) {
+	certificateStore, err := certificate.NewFileStore(
+		"kubelet-client",
+		certDirectory,
+		certDirectory,
+		"",
+		"")
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize certificate store: %v", err)
+	}
+	var rotation uint
+	if certRotation > 0 {
+		rotation = uint(certRotation)
+	}
+	clientCertificateManager, err := certificate.NewManager(&certificate.Config{
+		Template: &x509.CertificateRequest{
+			Subject: pkix.Name{
+				Organization: []string{"system:nodes"},
+				CommonName:   fmt.Sprintf("system:node:%s", nodeName),
+			},
+		},
+		Usages: []certificates.KeyUsage{
+			certificates.UsageDigitalSignature,
+			certificates.UsageKeyEncipherment,
+			certificates.UsageClientAuth,
+		},
+		CertificateStore:           certificateStore,
+		CertificateRotationPercent: rotation,
+		Rotation: func() {
+			glog.Info("The client certificate was rotated, restarting so changes take effect.")
+			os.Exit(0)
```

Comment:
This is too disruptive. My biggest concern about restarting the Kubelet is it would disrupt any open connections (watching logs, exec/attach/portforward connections). It would also stress the runtime, and may have other performance impacts.

Fortunately, the ListenAndServeTLS function is not too complicated, and should be pretty easy to unroll and swap in our own `Accept` implementation:

1. [`ListenAndServeTLS`](https://github.com/golang/go/blob/753452fac6f6963b5a6e38a239b05362385a3842/src/net/http/server.go#L2401) just sets up a tls listener, and calls `server.Serve(...)` on it.
2. [`tls.NewListener`](https://github.com/golang/go/blob/753452fac6f6963b5a6e38a239b05362385a3842/src/crypto/tls/tls.go#L63) just wraps a `tls.listener`, which simlpy wraps the `Conn` returned by the inner listener.

I think all we'd need to do is create our own listener wrapper which calls `tls.Server` with the `tls.Config` with the latest certs.